File: daemon/rflink/Controller.php
<?php


namespace Ething\RFLink;


use \Ething\RFLink\RFLink;
use \Ething\Device\RFLinkGateway;
use \Ething\Device\RFLinkNode;
use \Ething\Event;

abstract class Controller extends \Stream {
	
	const AUTOCONNECT_PERIOD = 60; // seconds	
	const RESPONSE_TIMEOUT = 10; // seconds	
	
	public $gateway = null;
	
	public $gatewayLibVersion = false;
	
	protected $lastAutoconnectLoop = 0;
	private $preventFailConnectLog = 0;
	
	private $logMessage = false;
	
	protected $isOpened = false;
	
	// response management
	private $responseListeners = array();
	
	public $options = array();
	
	public function __construct(RFLinkGateway $gateway, array $options = array()){
		
		$this->gateway = $gateway;
		$this->options = array_replace_recursive($this->options, $options);
	}
	
	public function ething(){
		return $this->gateway->ething;
	}
	
	public function name(){
		return $this->gateway->name();
	}
	
	public function __get($name){
		return isset($this->options[$name]) ? $this->options[$name] : null;
    }
	
	public function open(){
		$this->gateway->setConnectState(true);
		$this->isOpened = true;
		return true;
	}
	
	abstract public function read();
	abstract public function write($str);
	
	public function close(){
		$this->isOpened = false;
		$this->lastAutoconnectLoop = 0;
		$this->gateway->setConnectState(false);
		\Log::info("RFLink: closed");
		return true;
	}
	
	/*
	exemple of messages :
		20;00;Nodo RadioFrequencyLink - RFLink Gateway V1.1 - R46;
		20;01;MySensors=OFF;NO NRF24L01;
		20;02;setGPIO=ON;
		20;03;Cresta;ID=8301;WINDIR=0005;WINSP=0000;WINGS=0000;WINTMP=00c3;WINCHL=00c3;BAT=LOW;
		20;04;Cresta;ID=3001;TEMP=00b4;HUM=50;BAT=OK;
		20;05;Cresta;ID=2801;TEMP=00af;HUM=53;BAT=OK;
		20;06;NewKaku;ID=008440e6;SWITCH=1;CMD=OFF;
		20;02;VER=1.1;REV=46;BUILD=0c;
	*/
	function processLine($line){
		
		\Log::debug("RFLink: message received = {$line}");
		
		$gateway = $this->gateway;
		if($this->logMessage) $gateway->log($line);
		$inclusion = $gateway->inclusion;
		$gateway->updateSeenDate();
		
		$words = explode(';', rtrim($line,";"));
		$wordsCount = count($words);
		
		if($wordsCount<3) return;
		
		// keep only messages destined to the gateway
		if($words[0]!=="20") return;
		
		if($wordsCount === 3 || substr($words[2], 0, 4)==='VER='){
			// system command/response
			
			// does a user request wait for a response
			foreach($this->responseListeners as $i => $responseListener){
				// remove this item
				array_splice($this->responseListeners, $i, 1);
				
				if(is_callable($responseListener['callback']))
					call_user_func($responseListener['callback'], false, $line);
			}
			
			if(preg_match('/Nodo RadioFrequencyLink - RFLink Gateway V([\d\.]+) - R([\d]+)/', $words[2], $matches)){
				$gateway->set('version', $matches[1]);
				$gateway->set('revision', $matches[2]);
				\Log::info("RFLink: ver:{$matches[1]} rev:{$matches[2]}");
			} else if(preg_match('/;VER=([\d\.]+);REV=([\d]+);BUILD=([0-9a-fA-F]+);/', $line, $matches)) {
				$gateway->set('version', $matches[1]);
				$gateway->set('revision', $matches[2]);
				$gateway->set('build', $matches[3]);
				\Log::info("RFLink: ver:{$matches[1]} rev:{$matches[2]} build:{$matches[3]}");
			}
			
		} else {
			
			$protocol = $words[2];
			$args = array();
			
			for($i=3; $i<$wordsCount; $i++){
				
				if( $sepi = strpos($words[$i], '=') ){
					// key value pair
					$key = substr($words[$i], 0, $sepi);
					$value = substr($words[$i], $sepi+1);
					
					$args[$key] = $value;
				}
			}
			
			if(isset($args['ID'])){
				
				$device = null;
				
				$switchId = isset($args['SWITCH']) ? RFLink::convertSwitchId($args['SWITCH']) : null;
				
				$device = $gateway->getNode(array(
					'nodeId' => $args['ID'],
					'protocol' => $protocol,
					'switchId' => $switchId
				));
				
				if(!$device){
					if($inclusion){
						// the device does not exist !
						
						// find the best subType suited from the protocol and args
						if($subType = RFLink::getSubType($protocol, $args)){
						
							// create it !
							if($device = RFLinkNode::createDeviceFromMessage($subType, $protocol, $args, $gateway)){
								\Log::info("RFLink: new node ({$subType}) from {$line}");
							} else {
								\Log::error("RFLink: fail to create the node ({$subType}) from {$line}");
							}
						} else {
							\Log::warn("RFLink: unable to handle the message {$line}");
						}
					} else {
						\Log::warn("RFLink: new node from {$line}, rejected because inclusion=false");
					}
				}
				
				if($device){
					$device->processMessage($protocol, $args);
				}
			
			} else {
				\Log::warn("RFLink: unable to handle the message {$line}, no id.");
			}
			
		}
		
	}
	
	
	private $lastState_ = false;
	
	public function update(){
		// do some stuff regularly
		$now = microtime(true);
		
		// check for a deconnection
		if(!$this->isOpened && $this->isOpened != $this->lastState_)
			\Log::info("RFLink: disconnected");
		$this->lastState_ = $this->isOpened;
		
		// autoconnect
		if(!$this->isOpened && ($now - $this->lastAutoconnectLoop) > self::AUTOCONNECT_PERIOD ){
			try{
				$this->open();
				$this->preventFailConnectLog = 0;
			} catch(\Exception $e){
				$this->gateway->setConnectState(false);
				
				if($this->preventFailConnectLog % 20 === 0) \Log::warn("RFLink: unable to connect : {$e->getMessage()}");
				$this->preventFailConnectLog += 1;
			}
			$this->lastAutoconnectLoop = $now;
		}
		
		// check response timeout
		foreach($this->responseListeners as $i => $responseListener){
			if( ($now - $responseListener['ts']) > self::RESPONSE_TIMEOUT ){
				
				// remove this item
				array_splice($this->responseListeners, $i, 1);
				
				if(is_callable($responseListener['callback'])){
					call_user_func($responseListener['callback'], 'response timeout', null);
				}
			}
		}
	}
	
	
	/*
	*  $message message to send
	*  $callback (optional) function($error, $messageSent, $messageReceived = null)
	*  $waitResponse (optional) true|false wait for a response or not
	*/
	public function send($message, $callback = null, $waitResponse = null) {
		
		\Log::debug("RFLink: message send");
		
		if(!$this->isOpened){
			if(is_callable($callback)){
				call_user_func($callback, 'not connected', $message);
			}
			return 0;
		}
		
		
		$wb =  $this->write($message);
		
		if($waitResponse){
			// wait for a response
			$this->responseListeners[] = array(
				'callback' => function($error, $messageReceived) use($callback, $message) {
					if(is_callable($callback)) call_user_func($callback, $error, $message, $messageReceived);
				},
				'ts' => microtime(true),
				'messageSent' => $message
			);
		} else {
			if(is_callable($callback)) call_user_func($callback, false, $message);
		}
		
		return $wb;
	}
	
	protected $stream = null;
	
	public function getStreams(){
		return array($this->stream);
	}
	
	public function process($stream){
		$this->read();
	}
	
};

File: ething/rflink/Controller.py
# coding: utf-8

from future.utils import text_type, binary_type
from .helpers import convertSwitchId, getSubType
from .RFLinkNode import RFLinkNode
import re
import time


class Controller(object):
    
    AUTOCONNECT_PERIOD = 60; # seconds    
    RESPONSE_TIMEOUT = 10; # seconds    
    
    reset_attr = ['port', 'baudrate']
    
    def __init__ (self, gateway, transport):
        
        self._gateway = gateway
        self._ething = gateway.ething
        self._isOpened = False
        
        self._log = gateway.ething.log
        
        self._lastState = False
        self._lastAutoconnectLoop = 0
        self._preventFailConnectLog = 0
        
        # response management
        self._responseListeners = []
        
        # refresh the gateway each time the gateway properties changes
        self.ething.signalManager.bind('ResourceMetaUpdated', self.onResourceMetaUpdated)
        
        self.ething.scheduler.tick(self.update)
        
        self._transport = transport(self)
    
    
    def __del__(self):
        self.destroy()
    
    
    @property
    def gateway (self):
        return self._gateway
    
    @property
    def ething (self):
        return self._ething
    
    @property
    def log (self):
        return self._log
    
    @property
    def isOpened (self):
        return self._isOpened
    
    @property
    def transport (self):
        return self._transport
    
    def onResourceMetaUpdated(self, signal):
        if signal['resource'] == self.gateway.id and signal['rModifiedDate'] > self.gateway.modifiedDate:
            self.gateway.refresh()
            
            for attr in signal['attributes']:
                if attr in Controller.reset_attr:
                    self.restart()
                    break
    
    
    def restart(self):
        if self.isOpened:
            self.close()
            self.open()
    
    
    def destroy(self):
        self.close()
        self.ething.signalManager.unbind('ResourceMetaUpdated', self.onResourceMetaUpdated);
    
    def open (self):
        
        try:
            self.transport.open()
        except Exception as e:
            self.log.error(e)
            return False
        
        self._lastAutoconnectLoop = 0
        self.gateway.setConnectState(True)
        self._isOpened = True
        return True
    
    def close (self):
        
        self.transport.close()
        
        self._isOpened = False
        self.gateway.setConnectState(False)
        
        for responseListener in self._responseListeners:
            responseListener['callback']('disconnected', None);
        self._responseListeners = []
        
        return True
    
    

    # exemple of messages :
    #     20;00;Nodo RadioFrequencyLink - RFLink Gateway V1.1 - R46;
    #     20;01;MySensors=OFF;NO NRF24L01;
    #     20;02;setGPIO=ON;
    #     20;03;Cresta;ID=8301;WINDIR=0005;WINSP=0000;WINGS=0000;WINTMP=00c3;WINCHL=00c3;BAT=LOW;
    #     20;04;Cresta;ID=3001;TEMP=00b4;HUM=50;BAT=OK;
    #     20;05;Cresta;ID=2801;TEMP=00af;HUM=53;BAT=OK;
    #     20;06;NewKaku;ID=008440e6;SWITCH=1;CMD=OFF;
    #     20;02;VER=1.1;REV=46;BUILD=0c;
    
    def processLine (self, line):
        
        if isinstance(line, binary_type):
            line = line.decode('utf8')
        
        self.log.debug("RFLink: message received = %s" % line)
        
        with self.gateway as gateway:
            
            gateway.setConnectState(True)
            
            words = line.rstrip(';').split(';')
            wordsCount = len(words)
            
            if wordsCount<3:
                self.log.warn("RFLink: invalid message received '%s'" % line)
                return
            
            # keep only messages destined to the gateway
            if words[0] != "20":
                return
            
            
            if wordsCount == 3 or words[2][0:4] == 'VER=':
                # system command/response
                
                # does a user request wait for a response
                for responseListener in self._responseListeners:
                    responseListener['callback'](False, line);
                self._responseListeners = []
                
                matches = re.search('Nodo RadioFrequencyLink - RFLink Gateway V([\d\.]+) - R([\d]+)', words[2])
                if matches:
                    gateway._version = matches.group(1)
                    gateway._revision = matches.group(2)
                    self.log.info("RFLink: ver:%s rev:%s" % (matches.group(1), matches.group(2)))
                else:
                    matches = re.search(';VER=([\d\.]+);REV=([\d]+);BUILD=([0-9a-fA-F]+);', line)
                    if matches:
                        gateway._version = matches.group(1)
                        gateway._revision = matches.group(2)
                        gateway._build = matches.group(3)
                        self.log.info("RFLink: ver:%s rev:%s build:%s" % (matches.group(1), matches.group(2), matches.group(3)))
                    
                
                
            else:
                
                protocol = words[2]
                args = {}
                
                for i in range(3, wordsCount):
                    sepi = words[i].find('=')
                    if sepi >= 0:
                        # key value pair
                        key = words[i][0:sepi]
                        value = words[i][sepi+1:]
                        args[key] = value
                
                if 'ID' in args:
                    
                    switchId = convertSwitchId(args['SWITCH']) if 'SWITCH' in args else None
                    
                    device = gateway.getNode({
                        'nodeId' : args['ID'],
                        'protocol' : protocol,
                        'switchId' : switchId
                    })
                    
                    if not device:
                        if gateway.data.get('inclusion', False):
                            # the device does not exist !
                            
                            subType = getSubType(protocol, args)
                            
                            # find the best subType suited from the protocol and args
                            if subType:
                                
                                # create it !
                                device = RFLinkNode.createDeviceFromMessage(subType, protocol, args, gateway)
                                
                                if device:
                                    self.log.info("RFLink: new node (%s) from %s" % (subType, line))
                                else:
                                    self.log.error("RFLink: fail to create the node (%s) from %s" % (subType, line))
                                
                            else:
                                self.log.warn("RFLink: unable to handle the message %s" % line)
                            
                        else:
                            self.log.warn("RFLink: new node from %s, rejected because inclusion=False" % line)
                    
                    if device:
                        with device:
                            device.setConnectState(True)
                            device.processMessage(protocol, args)
                    
                
                else:
                    self.log.warn("RFLink: unable to handle the message %s, no id." % line)
            
            
        
        
    
    
    
    def update (self):
        # do some stuff regularly
        now = time.time()
        
        # check for a deconnection
        if not self.isOpened and self.isOpened != self._lastState:
            self.log.info("RFLink: disconnected")
        self._lastState = self.isOpened
        
        # autoconnect
        if not self.isOpened and (now - self._lastAutoconnectLoop) > Controller.AUTOCONNECT_PERIOD:
            try:
                self._lastAutoconnectLoop = now
                self.open()
                self._preventFailConnectLog = 0
            except Exception as e:
                
                if self._preventFailConnectLog % 20 == 0:
                    self.log.warn("RFLink: unable to connect : %s" % str(e))
                self._preventFailConnectLog += 1
        
        # check for timeout !
        i = 0
        while i < len(self._responseListeners):
            responseListener = self._responseListeners[i]
            
            if now - responseListener['ts'] > Controller.RESPONSE_TIMEOUT:
                
                # remove this item
                self._responseListeners.pop(i)
                i -= 1
                
                responseListener['callback']('response timeout', None)
            
            i += 1
            
    
    
    
    
    # $message message to send
    # $callback (optional) function(error, messageSent, messageReceived = None)
    # $waitResponse (optional) true|false wait for a response or not
    def send (self, message, callback = None, waitResponse = False):
        
        if isinstance(message, text_type):
            message = message.encode('utf-8')
        
        self.log.debug("RFLink: send message '%s'" % message)
        
        if not self.isOpened:
            self.log.warn("RFLink: unable to send message : not connected")
            if callable(callback):
                callback('not connected', message, None)
            return 0
        
        
        
        wb =  self.transport.write(message+b"\r\n")
        
        if waitResponse:
            
            def cb(error, messageReceived):
                if callable(callback):
                    callback(error, message, messageReceived)
            
            # wait for a response
            self._responseListeners.append({
                'callback' : cb,
                'ts' : time.time(),
                'messageSent' : message
            })
            
        else:
            if callable(callback):
                callback(False, message, None)
        
        return wb
    
        
    



